util/app/path/path.go

```diff
@@ -27,3 +29,59 @@ func Path(root, path string) (string, error) {
 	}
 	return appPath, nil
 }
+
+type OutOfBoundsSymlinkError struct {
+	File string
+	Err  error
+}
+
+func (e *OutOfBoundsSymlinkError) Error() string {
+	return "out of bounds symlink found"
+}
+
+// CheckOutOfBoundsSymlinks determines if basePath contains any symlinks that
+// are absolute or point to a path outside of the basePath. If found, an
+// OutOfBoundsSymlinkError is returned.
+func CheckOutOfBoundsSymlinks(basePath string) error {
+	absBasePath, err := filepath.Abs(basePath)
+	if err != nil {
+		return fmt.Errorf("failed to get absolute path: %v", err)
+	}
+	return filepath.Walk(absBasePath, func(path string, info os.FileInfo, err error) error {
+		if err != nil {
+			return fmt.Errorf("failed to walk for symlinks in %s: %v", absBasePath, err)
+		}
+		if files.IsSymlink(info) {
+			// We don't use filepath.EvalSymlinks because it fails without returning a path
+			// if the target doesn't exist.
+			linkTarget, err := os.Readlink(path)
+			if err != nil {
+				return fmt.Errorf("failed to read link %s: %v", path, err)
+			}
+			// get the path of the symlink relative to basePath, used for error description
+			linkRelPath, err := filepath.Rel(absBasePath, path)
+			if err != nil {
+				return fmt.Errorf("failed to get relative path for symlink: %v", err)
+			}
+			// deny all absolute symlinks
+			if filepath.IsAbs(linkTarget) {
+				return &OutOfBoundsSymlinkError{File: linkRelPath}
+			}
+			// get the parent directory of the symlink
+			dir := filepath.Dir(path)
+			// join the symlink target with the parent directory of the symlink
+			joined := filepath.Join(dir, linkTarget)
+			// resolve the relative path of the symlink target to the base path
+			rel, err := filepath.Rel(absBasePath, joined)
+			if err != nil {
+				return fmt.Errorf("failed to get relative path for symlink target: %v", err)
+			}
+			// if it's outside the repository, the path will contain ..
+			if strings.Contains(rel, string(os.PathSeparator)+"..") || strings.HasPrefix(rel, ".."+string(os.PathSeparator)) {
```

Comment:
I think `string(os.PathSeparator)+"..")` is too aggressive. `a/..bc` would match despite not being a traversal.

Reply:
Good catch again!

Come to think of it, `strings.HasPrefix(rel, ".."+string(os.PathSeparator)` should be sufficient. This is checking an already cleaned path and so anything outside of the repo should begin with `../`. Thanks to the implicit `filepath.Clean` from `filepath.Rel` It's not going to have any trickery like `./..` so we don't have to worry about those edge cases.

Reply:
Agreed, I think the `../` check is fine!